### ISSUE TITLE
Invalid capabilities in firstMatch element 0: a proxy type named MANUAL is specified, but proxy type must be 'pac', 'noproxy', 'autodetect', 'system', or 'manual'

### ISSUE THREAD

**Issue by kongkong56** [SeleniumHQ/selenium#4574]
## Meta -
OS:  WINDOWS10
Selenium Version:  3.5.1
Browser:  Internet Explorer 11

Browser Version:  11.1593.14393.0 64bit 

c# 
error :
Invalid capabilities in firstMatch element 0: a proxy type named MANUAL is specified, but proxy type must be 'pac', 'noproxy', 'autodetect', 'system', or 'manual'

**Comment by barancev** [comment]
For issues please provide a [concise reproducible test case](http://sscce.org/) and describe what results you are seeing and what results you expect.

See [CONTRIBUTING.md](https://github.com/SeleniumHQ/selenium/blob/master/CONTRIBUTING.md#issue-contributions)

**Comment by kongkong56** [comment]
Expected Behavior -

internet explorer browser should opened and can open  by proxy

Actual Behavior -

Unable to open internet explorer

Steps to reproduce -

Lunch hub and node with selenium standalone server 3.5.1
refer attached batch files to start hub and node

My code: 
  private InternetExplorerOptions GetOptions()
        {
            var options = new InternetExplorerOptions();
            var proxy = new OpenQA.Selenium.Proxy();
			options.UsePerProcessProxy = true;
			proxy.Kind = ProxyKind.Manual;
			proxy.HttpProxy = requestArgs.ProxyIP;
			proxy.SslProxy = requestArgs.ProxyIP;
			proxy.IsAutoDetect = false;
			proxy.NoProxy = string.Empty;
			options.Proxy = proxy;
			options.BrowserCommandLineArguments = "--private";
            options.IntroduceInstabilityByIgnoringProtectedModeSettings = true;
            options.IgnoreZoomLevel = true;
            return options;
        }
  private InternetExplorerDriverService InitService()
        {
            string driverPath = Helper.GetAppSetting("IedriverPath");
            var service = InternetExplorerDriverService.CreateDefaultService(driverPath);
            service.HideCommandPromptWindow = true;
            return service;

        }
                 var options = GetOptions();
                    var service = InitService();
                    driver = new InternetExplorerDriver(service, options);
                    driver.Manage().Window.Maximize();
                        driver.Navigate().GoToUrl(url);

Error:
Invalid capabilities in firstMatch element 0: a proxy type named MANUAL is specified, but proxy type must be 'pac', 'noproxy', 'autodetect', 'system', or 'manual' 

   在 OpenQA.Selenium.Remote.RemoteWebDriver.UnpackAndThrowOnError(Response errorResponse)
   在 OpenQA.Selenium.Remote.RemoteWebDriver.Execute(String driverCommandToExecute, Dictionary`2 parameters)
   在 OpenQA.Selenium.Remote.RemoteWebDriver.StartSession(ICapabilities desiredCapabilities)
   在 OpenQA.Selenium.Remote.RemoteWebDriver..ctor(ICommandExecutor commandExecutor, ICapabilities desiredCapabilities)
   在 OpenQA.Selenium.IE.InternetExplorerDriver..ctor(InternetExplorerDriverService service, InternetExplorerOptions options)
   在 DataAnalysis.SpiderService.Spider.IE.Get(String url, RequestArgs requestArgs)

**Comment by jimevans** [comment]
This is a bug in the IE driver. The spec language changed to indicate the passed-in value should be lowercased, and the driver does not adhere to that statement in the spec.

**Comment by barancev** [comment]
@jimevans Are you sure it's the driver issue? Would you clarify my reading of the specification. The spec lists proxy types in lowercase. IEDriverServer expects them to be in lowercase. But all language bindings send uppercase strings.

There was a patch that suggested to treat them as case-insensitive: https://github.com/w3c/webdriver/pull/870 , but I can't see it in the current editors draft of the specification, so it must be reverted.

**Comment by jimevans** [comment]
From [Section 7.1 (Proxy)](https://www.w3.org/TR/webdriver/#proxy), in the algorithm to "deserialize a proxy," step 3, substep 3:

> If key is equal to "proxyType", let value be the result of lowercasing value.

**Comment by barancev** [comment]
Editors draft have different wording:
https://w3c.github.io/webdriver/webdriver-spec.html#proxy

**Comment by jimevans** [comment]
You're right, of course, @barancev. I'm not in front of my laptop at the moment, and doing this on my phone, so I had an older draft of the spec queued up. Either way, it needs to be fixed in a component I maintain, so it's work for me either way.

**Comment by barancev** [comment]
@shs96c said the bindings should be changed to send proxy type in lowercase.

**Comment by barancev** [comment]
Java binding sends wrong payload too: https://gist.github.com/barancev/7ab7bbe91a59ab7bcbfb2c399d0202ab

**Comment by barancev** [comment]
Sorry, Java appears to be all right, the parts of payload required by W3C spec contain lowercased proxy type.

**Comment by shs96c** [comment]
Yeah. This is messy. We want to retain backwards compatibility, so "`desiredCapabilities`" should retain the upper-casing for `proxyType` (meaning a new client can talk to a 2.x server).

The spec, however, says anything in the w3c new session payload must be spec compliant, which means that "`capabilities.firstMatch`" and "`capabilities.alwaysMatch`" should have a lower-case `proxyType`.

**Comment by sandramedina** [comment]
Any update on this issue? I'm getting the same error, and here is my code:

```
Server server = new Server(@"C:\Users\sandra\Desktop\BrowserEfficiencyTest\BrowserEfficiencyTest\Dependencies\BrowserMobProxy\bin\browsermob-proxy.bat");
server.Start();
Client client = server.CreateProxy();
client.NewHar("google");
var seleniumProxy = new Proxy { HttpProxy = client.SeleniumProxy, FtpProxy = client.SeleniumProxy, SslProxy = client.SeleniumProxy };

var options = new InternetExplorerOptions();
options.Proxy = seleniumProxy;
            
InternetExplorerDriver driver = new InternetExplorerDriver(options);
driver.Navigate().GoToUrl("http://www.google.co.uk");
```

**Comment by p0deje** [comment]
91be3307526015b641c830f885103f5253829446 fixes this in Ruby bindings.

**Comment by jimevans** [comment]
Fixed for .NET in eead3f1

**Comment by lmtierney** [comment]
Fixed for python in e9d3ff1

**Comment by jleyba** [comment]
nodejs bindings have always sent lowercase proxy type.

Closing this issue since node was the last language tagged.
